drivers/pppou.c

```diff
@@ -23,109 +22,199 @@
+#define log_warn(fmt, ...)
+#define log_error(fmt, ...)
+
 #endif
 
-#define PPPOS_READ_AT_TIMEOUT_STEP_MS 		5
-#define PPPOS_READ_DATA_TIMEOUT_STEP_MS 	10
+#define PPPOU_READ_DATA_TIMEOUT_STEP_MS  10
+
+#define PPPOU_TRYOPEN_SERIALDEV_SEC      3
+#define PPPOU_CONNECT_RETRY_SEC          5
 
-#define PPPOS_TRYOPEN_SERIALDEV_SEC 		3
-#define PPPOS_CONNECT_RETRY_SEC 		5
-#define PPPOS_CONNECT_CMD_RETRY_MS		500
 
 /****** serial handling ******/
 
-static void serial_close(int fd)
+
+struct serial_speed_s {
+	const char *str;
+	speed_t speed;
+};
+
+
+static int _speedcmp(const void *l, const void *r)
 {
-	log_info("close()");
+	return strcmp(((struct serial_speed_s *)l)->str, ((struct serial_speed_s *)r)->str);
+}
 
-	if (fd >= 0)
-		close(fd);
 
-//	state->fd = -1;
-	// NOTE: set DISCONNECTED in status callback
-	// state->conn_state = CONN_STATE_DISCONNECTED;
+static speed_t serial_speed_from_string(const char *str)
+{
+	/* table is lexically sorted by `str` */
+	static const struct serial_speed_s stab[] = {
+		{ "0", B0 },           { "110", B110 },       { "115200", B115200 },
+		{ "1200", B1200 },     { "134", B134 },       { "150", B150 },
+		{ "1800", B1800 },     { "19200", B19200 },   { "200", B200 },
+		{ "230400", B230400 }, { "2400", B2400 },     { "300", B300 },
+		{ "38400", B38400 },   { "460800", B460800 }, { "4800", B4800 },
+		{ "50", B50 },         { "57600", B57600 },   { "600", B600 },
+		{ "75", B75 },         { "9600", B9600 },
+	};
+
+	struct serial_speed_s *sptr;
+
+	if (str && (sptr = bsearch(&str, stab, sizeof(stab) / sizeof(stab[0]), sizeof(stab[0]), _speedcmp)))
+		return sptr->speed;
+
+	return -1;
 }
 
-static void serial_set_non_blocking(int fd)
+
+static int serial_set_blocking(int fd, int set_block)
 {
-	int flags = fcntl(fd, F_GETFL, 0);
-	if (fcntl(fd, F_SETFL, flags | O_NONBLOCK) < 0)
-		log_error("%s() : fcntl(%d, O_NONBLOCK) = (%d -> %s)", __func__, fd, errno, strerror(errno));
+	int flags;
+
+	if ((flags = fcntl(fd, F_GETFL, 0)) >= 0) {
+		if (set_block)
+			flags = fcntl(fd, F_SETFL, flags & ~O_NONBLOCK);
+		else
+			flags = fcntl(fd, F_SETFL, flags | O_NONBLOCK);
+	}
+
+	if (flags < 0) {
+		log_error("%s(): fcntl(%d, O_%sBLOCK) = (%d -> %s)", __func__, fd, set_block ? "" : "NON", errno, strerror(errno));
+	}
+
+	return flags;
 }
 
-#if 0
-static void serial_set_blocking(int fd)
+
+static int serial_open(const char *devname, speed_t speed)
 {
-	int flags = fcntl(fd, F_GETFL, 0);
-	if (fcntl(fd, F_SETFL, flags & ~O_NONBLOCK) < 0)
-		log_error("%s() : fcntl(%d, ~O_NONBLOCK) = (%d -> %s)", __func__, fd, errno, strerror(errno));
+	oid_t oid;
+	int fd, ret, cnt;
+	struct termios tio;
+
+	if (!devname)
+		return -1;
+
+	/* try if uart is registered */
+	for (cnt = 0; (ret = lookup(devname, NULL, &oid)) < 0; cnt++) {
+		usleep(200 * 1000);
+		if (cnt > 3)
+			return ret;
+	}
+
+	if ((fd = open(devname, O_RDWR | O_NOCTTY | O_NONBLOCK)) < 0)
+		return fd;
+
+	if ((ret = serial_set_blocking(fd, 0)) < 0)
```

Comment:
Why setting `O_NONBLOCK` twice? (it was already set with `open` flags)?

Reply:
That's right, because serial_set_blocking is a static function and only here it is used I will remove a call and the definition. push-force the patch.